pkg/controllers/management/clusterupstreamrefresher/cluster_upstream_refresher.go

```diff
@@ -59,39 +59,24 @@ func (c *clusterRefreshController) onClusterChange(key string, cluster *mgmtv3.C
 		// not a KEv2 cluster
 		return cluster, nil
 	}
-
 	if !ready {
-		return cluster, nil
-	}
-
-	var lastRefreshTime string
-	var err error
-
-	if cluster.Annotations != nil {
-		lastRefreshTime = cluster.Annotations[clusterLastRefreshTime]
-	}
-
-	providerRefreshInterval, err := getProviderRefreshInterval(provider)
-	if err != nil {
-		return cluster, err
+		return c.updateCluster(cluster)
 	}
 
-	shouldRefresh, err := shouldRefreshCluster(providerRefreshInterval, lastRefreshTime)
+	providerRefreshInterval := getProviderRefreshInterval(provider)
+	shouldRefresh, err := shouldRefreshCluster(providerRefreshInterval, cluster.Annotations[clusterLastRefreshTime])
 	if err != nil {
 		return cluster, err
 	}
 
 	if shouldRefresh {
-		return cluster, err
-	}
-
-	cluster, err = c.refreshClusterUpstreamSpec(cluster, provider)
-	if err != nil {
-		return cluster, err
+		cluster, err = c.refreshClusterUpstreamSpec(cluster, provider)
+		if err != nil {
+			return cluster, err
+		}
 	}
 
 	c.clusterEnqueueAfter(key, providerRefreshInterval)
```

Comment:
In my testing, if the cluster object is updated before a refresh is due to happen and we don't re-enqueue then the cluster update will cancel any previous re-enqueue. In other words, in this situation, the cluster refresh won't happen again until the next cluster update.

I also checked the code of the worker queue: if there is already a re-enqueue for a cluster in the queue and we make a call to enqueue it again, it will take the re-enqueue that will put the cluster through the controllers the soonest.

I hope that these are consistent with your understanding of how things work.

Reply:
Verified the following scenario if `c.clusterEnqueueAfter(key, providerRefreshInterval)` is moved inside the if-block:
1. Start an update of the cluster that completes between sync calls (adding a node will do this if timed correctly).
2. The cluster is not successfully re-queued by the refresher. That is, refreshing stops until the cluster is updated for some reason.

Reply:
@thedadams thanks for doing a deeper dive on that, it's all valuable info.